runtime/util/hshelp.c

```diff
@@ -3046,6 +3046,33 @@ verifyClassesAreCompatible(J9VMThread * currentThread, jint class_count, J9JVMTI
 			return rc;
 		}
 
+#if defined(J9VM_OPT_VALHALLA_NESTMATES)
+		/* Verify that the nestmates attributes - nest host & nest members - are the same */
+
+		if (J9ROMCLASS_NESTHOSTNAME(originalROMClass) != J9ROMCLASS_NESTHOSTNAME(replacementROMClass)) {
+			return JVMTI_ERROR_UNSUPPORTED_REDEFINITION_CLASS_ATTRIBUTE_CHANGED;
+		}
+
+		{
+			U_16 nestMemberCount = originalROMClass->nestMemberCount;
+			if (nestMemberCount != replacementROMClass->nestMemberCount) {
+				return JVMTI_ERROR_UNSUPPORTED_REDEFINITION_CLASS_ATTRIBUTE_CHANGED;
+			} else if (0 != nestMemberCount) {
+				J9SRP *originalNestMembers = J9ROMCLASS_NESTMEMBERS(originalROMClass);
+				J9SRP *replacementNestMembers = J9ROMCLASS_NESTMEMBERS(replacementROMClass);
+				U_16 i = 0;
+
+				for (i = 0; i < nestMemberCount; i++) {
+					J9UTF8 *originalNestMemberName = NNSRP_GET(originalNestMembers[i], J9UTF8*);
+					J9UTF8 *replacementNestMemberName = NNSRP_GET(replacementNestMembers[i], J9UTF8*);
+					if (!J9UTF8_EQUALS(originalNestMemberName, replacementNestMemberName)) {
+						return JVMTI_ERROR_UNSUPPORTED_REDEFINITION_CLASS_ATTRIBUTE_CHANGED;
```

Comment:
This loop enforces ordering of nest members - is reodering not allowed?

Reply:
We do the same thing for implemented interfaces (which may also not be correct).

Reply:
Changing the NestHost or NestMembers class attributes is disallowed in the current spec draft.

In the version that I've been working from, it writes:

```
Unless the canUnrestrictedlyRedefineClasses capability is present the following redefinitions are restricted:
    changing the schema (the fields)
    changing the hierarchy (superclasses, interfaces)
    deleting a method
    changing class modifiers
    changing method modifiers
    changing the NestHost or NestMembers class attributes
```
(found under 'RedefineClasses Command (18)' here: http://cr.openjdk.java.net/~dholmes/8010319/specs/JDWP/diff.html)

My interpretation of this is that re-ordering the nest members counts as modifying them

Reply:
OK